### PR TITLE
Make numFeasibleNodesToFind stateless

### PR DESCRIPTION
**Pull request by Jerry-Ge**
Signed-off-by: Jerry Ge <jerry.ge@arm.com>

#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
This PR is the second part of issue #101135 which will Unroll ScheduleAlgorithm#numFeasibleNodesToFind() to be stateless for future code refactor. 

#### Which issue(s) this PR fixes:

Part of #101135

#### Special notes for your reviewer:
NONE

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by Jerry-Ge** [comment]
/cc @Huang-Wei  @alculquicondor  @ahg-g  Hi guys, it would be really appreciated if you can review this PR which belongs the original #101135. I will squash the commits once the it's ready to be merged. :D 

Thanks, 

Jerry

File: test/integration/scheduler/queue_test.go [hunk 1]
```diff
@@ -137,7 +137,7 @@ func TestServiceAffinityEnqueue(t *testing.T) {
 		t.Fatalf("Cannot find the profile for Pod %v", podInfo.Pod.Name)
 	}
 	// Schedule the Pod manually.
-	_, fitError := testCtx.Scheduler.Algorithm.Schedule(ctx, nil, fwk, framework.NewCycleState(), podInfo.Pod)
+	_, fitError := testCtx.Scheduler.Algorithm.Schedule(30, ctx, nil, fwk, framework.NewCycleState(), podInfo.Pod)
```
Comment (chendave) [inline test/integration/scheduler/queue_test.go:140]: maybe a const here? fine as is since this is just a test.

**Review by chendave — COMMENTED** [review]
/lgtm

**Comment by alculquicondor** [comment]
I'm not following, why do you need this function stateless? What is the next step?

**Comment by Jerry-Ge** [comment]
> I'm not following, why do you need this function stateless? What is the next step?

Hi, in the first step I unroll the extender function to make the related member functions to be stateless from genericScheduler. So, I am planning to unroll the remaining members functions to be stateless and then migrate to Scheduler.go in the final steps. Does this sound reasonable or are there better strategies? Thanks for the comments😊

Jerry

**Comment by alculquicondor** [comment]
If you are going to end up unrolling every function, then we will have to roll them back when we move everything back to the pkg/scheduler package.

Maybe we should just go ahead and merge the packages now. @Huang-Wei ?

**Comment by Jerry-Ge** [comment]
> If you are going to end up unrolling every function, then we will have to roll them back when we move everything back to the pkg/scheduler package.
> 
> Maybe we should just go ahead and merge the packages now. @Huang-Wei ?

Yes, that's true. Let me know what do you guys think are there any better strategies? Thanks,

File: pkg/scheduler/core/generic_scheduler.go [hunk 2]
```diff
@@ -90,7 +89,7 @@ func (g *genericScheduler) snapshot() error {
 // Schedule tries to schedule the given pod to one of the nodes in the node list.
 // If it succeeds, it will return the name of the node.
 // If it fails, it will return a FitError error with reasons.
-func (g *genericScheduler) Schedule(ctx context.Context, extenders []framework.Extender, fwk framework.Framework, state *framework.CycleState, pod *v1.Pod) (result ScheduleResult, err error) {
+func (g *genericScheduler) Schedule(percentageOfNodesToScore int32, ctx context.Context, extenders []framework.Extender, fwk framework.Framework, state *framework.CycleState, pod *v1.Pod) (result ScheduleResult, err error) {
```
Comment (Huang-Wei) [inline pkg/scheduler/core/generic_scheduler.go:92]: Keep `context` as the first parameter: https://github.com/golang/go/wiki/CodeReviewComments#contexts

Nit: we can shorten `percentageOfNodesToScore` to `pctOfNodesToScore`, but up to you :)

File: pkg/scheduler/core/generic_scheduler.go [hunk 3]
```diff
@@ -168,12 +167,12 @@ func (g *genericScheduler) selectHost(nodeScoreList framework.NodeScoreList) (st
 
 // numFeasibleNodesToFind returns the number of feasible nodes that once found, the scheduler stops
 // its search for more feasible nodes.
-func (g *genericScheduler) numFeasibleNodesToFind(numAllNodes int32) (numNodes int32) {
-	if numAllNodes < minFeasibleNodesToFind || g.percentageOfNodesToScore >= 100 {
+func numFeasibleNodesToFind(percentageOfNodesToScore int32, numAllNodes int32) (numNodes int32) {
```
Comment (Huang-Wei) [inline pkg/scheduler/core/generic_scheduler.go:170]: nit:

```suggestion
func numFeasibleNodesToFind(percentageOfNodesToScore, numAllNodes int32) (numNodes int32) {
```

File: pkg/scheduler/core/generic_scheduler.go [hunk 4]
```diff
@@ -190,14 +189,14 @@ func (g *genericScheduler) numFeasibleNodesToFind(numAllNodes int32) (numNodes i
 	return numNodes
 }
 
-func (g *genericScheduler) evaluateNominatedNode(ctx context.Context, extenders []framework.Extender, pod *v1.Pod, fwk framework.Framework, state *framework.CycleState, diagnosis framework.Diagnosis) ([]*v1.Node, error) {
+func (g *genericScheduler) evaluateNominatedNode(percentageOfNodesToScore int32, ctx context.Context, extenders []framework.Extender, pod *v1.Pod, fwk framework.Framework, state *framework.CycleState, diagnosis framework.Diagnosis) ([]*v1.Node, error) {
```
Comment (Huang-Wei) [inline pkg/scheduler/core/generic_scheduler.go:192]: Ditto: ensure context as the first parameter.

File: pkg/scheduler/scheduler.go [hunk 5]
```diff
@@ -73,6 +73,8 @@ type Scheduler struct {
 
 	Extenders []framework.Extender
 
+	PercentageOfNodesToScore int32
```
Comment (Huang-Wei) [inline pkg/scheduler/scheduler.go:76]: Add a comment here.

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -508,7 +510,7 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 	state.SetRecordPluginMetrics(rand.Intn(100) < pluginMetricsSamplePercent)
 	schedulingCycleCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	scheduleResult, err := sched.Algorithm.Schedule(schedulingCycleCtx, sched.Extenders, fwk, state, pod)
+	scheduleResult, err := sched.Algorithm.Schedule(sched.PercentageOfNodesToScore, schedulingCycleCtx, sched.Extenders, fwk, state, pod)
```
Comment (Huang-Wei) [inline pkg/scheduler/scheduler.go:513]: Ditto: context as the first parameter.

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 7]
```diff
@@ -1434,12 +1429,12 @@ func TestFairEvaluationForNodes(t *testing.T) {
 	}
 
 	// To make numAllNodes % nodesToFind != 0
-	g.percentageOfNodesToScore = 30
-	nodesToFind := int(g.numFeasibleNodesToFind(int32(numAllNodes)))
+	percentageOfNodesToScore := int32(30)
```
Comment (Huang-Wei) [inline pkg/scheduler/core/generic_scheduler_test.go:1432]: nit: I think `var xyz int32 = 30` is more conventional.

**Review by Huang-Wei — COMMENTED** [review]
Some nits.

**Comment by Huang-Wei** [comment]
> /cc @Huang-Wei any comments? :)

Just back from vacation, sry for the delay.

**Comment by alculquicondor** [comment]
What about skipping this step and merging the packages?

**Comment by Huang-Wei** [comment]
> What about skipping this step and merging the packages?

Technically I'm fine with either approach. But given the extenders stuff has been refactored out, and resting stuff doesn't seem that functionally different and thus doesn't need to separate into multiple PRs. So yes, merging the packages in one PR is fine with me.

> If you are going to end up unrolling every function, then we will have to roll them back when we move everything back to the pkg/scheduler package.

This is also a valid point.

**Comment by Jerry-Ge** [comment]
Thanks Guys 😎

So, I will close this one and start to merge the packages first in the next PR.

**Comment by alculquicondor** [comment]
You can start by copying the `generic_sheduler.go` file to the `pkg/scheduler` folder and just fixing the imports.
Next, you can make all the functions in `generic_scheduler.go` be members of the `Scheduler` struct.

You can make the above 2 in 2 commits in the same PR.

Next, you can merge whatever functions make sense to be a single function (this one is open for debate later).

**Comment by Huang-Wei** [comment]
@Jerry-Ge are you still working on this?

**Pull request closed without merging**